### DISCUSSION TITLE
How to decrease Dockerfile build size?

### DISCUSSION THREAD

**Discussion by max10rogerio** [vercel/next.js#61767] [Help] [answered]
### Summary

I have a simple project with:
- nextjs 14
- chakra ui
- sentry

After building the Dockerfile, that's create a image with ~988mb

Analising the content of image, I have:
- node18-alpine: 166mb
- .next: 254mb
- node_modules: 470mb

### `.next` folder:
Inside this folder, after the build, these folders were created
![image](https://github.com/vercel/next.js/assets/21085292/6d9c457c-6017-4f8e-90f1-02b5f8eab9af)

In the cache folder there is a webpack folder which size is 241mb
![image](https://github.com/vercel/next.js/assets/21085292/173dccb6-bd65-4ad8-979d-8e60ca9f51ad)
Why the client folder is this big if I have one "hello" page in my project? How to optimize it?

### `node_modules` folder:
Inside the node_modules folder, I have 4 large packages, 2 are from nextjs (353mb), why? How to optimize it?

![image](https://github.com/vercel/next.js/assets/21085292/b23bff0e-8629-4c9f-9b40-9f234e68ef8e)

Only the Next.js files are near 594MB (~60% of the total size).

next.config.mjs
```js
import { withSentryConfig } from "@sentry/nextjs";

/** @type {import('next').NextConfig} */
const nextConfig = {};

export default withSentryConfig(
  nextConfig,
  {
    // For all available options, see:
    // https://github.com/getsentry/sentry-webpack-plugin#options

    // Suppresses source map uploading logs during build
    silent: true,
    org: process.env.NEXT_PUBLIC_SENTRY_ORG,
    project: process.env.NEXT_PUBLIC_SENTRY_PROJECT,
    url: process.env.NEXT_PUBLIC_SENTRY_URL,
    authToken: process.env.NEXT_PUBLIC_SENTRY_AUTH_TOKEN,
  },
  {
    // For all available options, see:
    // https://docs.sentry.io/platforms/javascript/guides/nextjs/manual-setup/

    // Upload a larger set of source maps for prettier stack traces (increases build time)
    widenClientFileUpload: true,

    // Transpiles SDK to be compatible with IE11 (increases bundle size)
    transpileClientSDK: true,

    // Routes browser requests to Sentry through a Next.js rewrite to circumvent ad-blockers (increases server load)
    tunnelRoute: "/monitoring",

    // Hides source maps from generated client bundles
    hideSourceMaps: true,

    // Automatically tree-shake Sentry logger statements to reduce bundle size
    disableLogger: true,
    automaticVercelMonitors: false,
  }
);
```

Dockerfile
```Dockerfile
FROM node:18-alpine AS base

# 1. Install dependencies only when needed
FROM base AS deps
# Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
RUN apk add --no-cache libc6-compat

WORKDIR /app

# Install dependencies based on the preferred package manager
COPY package.json yarn.lock* package-lock.json* pnpm-lock.yaml* ./
RUN \
  if [ -f yarn.lock ]; then yarn --frozen-lockfile; \
  elif [ -f package-lock.json ]; then npm ci; \
  elif [ -f pnpm-lock.yaml ]; then corepack enable pnpm && pnpm i; \
  else echo "Lockfile not found." && exit 1; \
  fi


# 2. Rebuild the source code only when needed
FROM base AS builder
WORKDIR /app
COPY --from=deps /app/node_modules ./node_modules
COPY . .

ENV NEXT_TELEMETRY_DISABLED=1

RUN npm run build

# 3. Production image, copy all the files and run next
FROM base AS runner
WORKDIR /app

ENV NODE_ENV production
# Uncomment the following line in case you want to disable telemetry during runtime.
ENV NEXT_TELEMETRY_DISABLED 1
ENV SENTRY_ENVIRONMENT production

RUN addgroup -g 1001 -S nodejs
RUN adduser -S nextjs -u 1001

COPY --from=builder /app/public ./public

# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=builder /app/node_modules ./node_modules
COPY --from=builder /app/package.json ./package.json
COPY --from=builder --chown=nextjs:nodejs /app/.next ./.next
COPY --from=builder --chown=nextjs:nodejs /app/next.config.mjs ./

# Change prepare script to does not install husky in production
RUN npm pkg set scripts.prepare="echo \"Production does not need husky, skipping installation\"" \
  && npm install --omit=dev

USER nextjs

EXPOSE 3000

ENV PORT 3000
ENV HOSTNAME localhost

CMD ["npm", "start"]
```

### Additional information

```js
Operating System:
  Platform: linux
  Arch: x64
  Version: #1 SMP Thu Oct 5 21:02:42 UTC 2023
Binaries:
  Node: 18.19.0
  npm: 10.2.3
  Yarn: 1.22.21
  pnpm: N/A
Relevant Packages:
  next: 14.1.0
  eslint-config-next: 14.1.0
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

**Comment by GabenGar** [comment]
> Inside the node_modules folder, I have 4 large packages, 2 are from nextjs (353mb), why?

Where do you think the code for handling your templates comes from, considering NextJS is a swiss army knife of web servers? Since base OS + dependencies + your code is a baseline for any docker image, there is nothing to do to "reduce" it further, as far as prod image concerned. So first you have to formulate why do you need the size reduction in the first place.

**Reply by max10rogerio** [reply]
I just want to know if there are any additional practices to further optimize the final Docker image. I understand that Next.js comes with various functionalities, and that impacts the final image size. I was just surprised that a simple project resulted in a 1GB image.

**Reply by icyJoseph** [reply, marked as answer]
Hi,

I think you can try to use standalone mode, which further optimizes the build step. 

- https://nextjs.org/docs/app/api-reference/next-config-js/output#automatically-copying-traced-files

> This feature helps reduce the size of deployments drastically. Previously, when deploying with Docker you would need to have all files from your package's dependencies installed to run next start. Starting with Next.js 12, you can leverage Output File Tracing in the .next/ directory to only include the necessary files.

**Reply by max10rogerio** [reply]
Hi,

Thanks for your reply.

My Docker image size is now 175MB using standalone config. I added Sharp, and the image size went up to 211MB – that's nice!

Thanks again!

**Reply by icyJoseph** [reply]
No worries mate! Please don't forget to mark as answered :)
